core/src/main/java/org/apache/iceberg/view/PropertiesUpdate.java

```diff
@@ -0,0 +1,112 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.view;
+
+import static org.apache.iceberg.TableProperties.COMMIT_MAX_RETRY_WAIT_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_MAX_RETRY_WAIT_MS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_MIN_RETRY_WAIT_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_MIN_RETRY_WAIT_MS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_RETRIES;
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_RETRIES_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT;
+
+import java.util.Map;
+import java.util.Set;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.Tasks;
+
+class PropertiesUpdate implements UpdateViewProperties {
+  private final ViewOperations ops;
+  private final Map<String, String> updates = Maps.newHashMap();
+  private final Set<String> removals = Sets.newHashSet();
+  private ViewMetadata base;
+
+  PropertiesUpdate(ViewOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+  }
+
+  @Override
+  public Map<String, String> apply() {
+    this.base = ops.refresh();
+
+    Map<String, String> newProperties = Maps.newHashMap();
+    for (Map.Entry<String, String> entry : base.properties().entrySet()) {
+      if (!removals.contains(entry.getKey())) {
+        newProperties.put(entry.getKey(), entry.getValue());
+      }
+    }
+
+    newProperties.putAll(updates);
+    return newProperties;
+  }
+
+  @Override
+  public void commit() {
+    Tasks.foreach(ops)
+        .retry(
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_NUM_RETRIES, COMMIT_NUM_RETRIES_DEFAULT))
+        .exponentialBackoff(
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_MIN_RETRY_WAIT_MS, COMMIT_MIN_RETRY_WAIT_MS_DEFAULT),
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_MAX_RETRY_WAIT_MS, COMMIT_MAX_RETRY_WAIT_MS_DEFAULT),
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_TOTAL_RETRY_TIME_MS, COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT),
+            2.0 /* exponential */)
+        .onlyRetryOn(CommitFailedException.class)
+        .run(
+            taskOps -> {
+              Map<String, String> newProperties = apply();
```

Comment:
I don't think this should use `apply` because it is calling `setProperties` and `removeProperties` directly. If you call `setProperties` with `newProperties` then `ViewMetadata` is responsible for merging the properties.

I'd probably update `apply` to run some `internalApply(ViewMetadata)` that produces `ViewMetadata`. Then `apply` can call that and return `properties` and this can also call it to build metadata for the commit.

Reply:
you're right. This is code that was aligned with the old builder. I've updated this to how you suggested it